packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -127,34 +127,37 @@ contract MerkleDistributor is Ownable {
      *
      ****************************/
 
-    // Batch claims for a reward currency for an account to save gas. We only allow
-    // batching the same reward token and the same account because this allows us to effect
-    // the most gas optimizations for the user by precomputing the total amount of the
-    // chosen reward currency to send to the user in a single transfer transaction.
-    // If we allowed multiple accounts or multiple reward currencies,
-    // then this function would still reduce to multiple ERC20.transfer calls.
-    function claimMulti(
-        Claim[] memory claims,
-        address rewardToken,
-        address account
-    ) external {
-        uint256 amountToClaim = 0;
-        for (uint256 i = 0; i < claims.length; i++) {
-            Claim memory claim = claims[i];
-            require(claim.account == account, "Invalid account in batch claim");
-            address _rewardToken = address(merkleWindows[claim.windowIndex].rewardToken);
-            require(_rewardToken == rewardToken, "Invalid rewardToken in batch claim");
-            _verifyAndMarkClaimed(claim);
-            amountToClaim = amountToClaim.add(claim.amount);
+    // Batch claims to reduce gas versus individual submitting all claims. Optimistically tries to batch
+    // together consecutive claims for the same account and same reward token to reduce gas.
+    function claimMulti(Claim[] memory claims) external {
+        uint256 batchedAmount = 0;
+        uint256 claimCount = claims.length;
+        for (uint256 i = 0; i < claimCount; i++) {
+            Claim memory _claim = claims[i];
+            _verifyAndMarkClaimed(_claim);
+            batchedAmount = batchedAmount.add(_claim.amount);
+
+            // If the next claim is NOT the same account or the same token (or this claim is the last one),
+            // then disburse the `batchedAmount` to the current claim's account for the current claim's reward token.
+            uint256 nextI = i.add(1);
+            address currentRewardToken = address(merkleWindows[_claim.windowIndex].rewardToken);
+            if (
+                nextI == claimCount || // This claim is last claim.
+                address(merkleWindows[claims[nextI].windowIndex].rewardToken) != currentRewardToken || // Next claim reward token is different than current one.
+                claims[nextI].account != _claim.account // Next claim account is different than current one.
+            ) {
+                uint256 amountToDisburse = batchedAmount;
+                batchedAmount = 0;
+                IERC20(currentRewardToken).safeTransfer(_claim.account, amountToDisburse);
```

Comment:
Why can't we do something like this?
```suggestion
                IERC20(currentRewardToken).safeTransfer(_claim.account, batchedAmount);
                batchedAmount = 0;
```

Reply:
I'd rather the external call come last for reentrancy reasons

Reply:
I don't think re-entrancy can affect this variable because batchedAmount is an in-memory variable that's local to this level in the stack. i.e. if you re-entered, batchedAmount would be pointing to a different spot in memory and couldn't interfere with the original call.

But I also get being cautious around re-entrancy. Up to you.

Reply:
I'll test the change you proposed for gas changes